File: fortran_version/ran2.f90
FUNCTION RAN2(idum)

  IMPLICIT NONE
  INTEGER, PARAMETER :: im1=2147483563,im2=2147483399,imm1=im1-1,&
       ia1=40014,ia2=40692,iq1=53668,iq2=52774,ir1=12211,&
       ir2=3791,ntab=32,ndiv=1+imm1/ntab
  DOUBLE PRECISION , PARAMETER ::   am=1.d0/im1,eps=1.d-14,rnmx=1.d0-eps
  DOUBLE PRECISION :: ran2
  INTEGER, DIMENSION(ntab) :: iv
  INTEGER :: idum,idum2,j,k,iy

  save iv,iy,idum2
  data idum2/123456789/,iv /ntab*0/,iy /0/

  if(idum.le.0) then
     idum=max(-idum,1)
     idum2=idum
     do j=ntab+8,1,-1
        k=idum/iq1
        idum=ia1*(idum-k*iq1)-ir1*k
        if(idum.lt.0) idum=idum+im1
        if(j.le.ntab) iv(j)=idum
     end do
     iy=iv(1)
  endif

  k=idum/iq1
  idum=ia1*(idum-k*iq1)-ir1*k
  if(idum.lt.0) idum=idum+im1
  k=idum2/iq2
  idum2=ia2*(idum2-k*iq2)-ir2*k

  if(idum2.lt.0) idum2=idum2+im2

  j=1+iy/ndiv
  iy=iv(j)-idum2
  iv(j)=idum
  if (iy.lt.1)iy=iy+imm1
  ran2=min(am*iy,rnmx)

END FUNCTION RAN2

File: cpp_version/ran2.cpp
#include <math.h>

using namespace std;

#define IM1 2147483563
#define IM2 2147483399
#define AM (1.0/IM1)
#define IMM1 (IM1-1)
#define IA1 40014
#define IA2 40692
#define IQ1 53668
#define IQ2 52774
#define IR1 12211
#define IR2 3791
#define NTAB 32
#define NDIV (1+IMM1/NTAB)
#define EPS 1.2e-7
#define RNMX (1.0-EPS)

long SEED = -1973;

//
//  Paul de Bakker, 6 Aug 2000                             
//
//  FROM NUMERICAL RECIPES IN C
//
//  Long period (>2 E18) random number generator of
//  L'Ecuyer with Bays-Durham suffle and added
//  safeguards. Returns a uniform random deviate between
//  0.0 and 1.0 (exclusive of the endpoint values).
//  Call with idum a negative integer to initialise;
//  thereafter, do not alter idum between successive
//  deviates in a sequence. RNMX should approximate the
//  largest floating value that is less than 1.
//

double ran2(long *idum)
{
  int j;
  long k;
  static long idum2 = 123456789;
  static long iy = 0;
  static long iv[NTAB];
  double temp;

  if (*idum <= 0) {
    if (-(*idum) < 1) *idum = 1;
    else *idum = -(*idum);
    idum2 = (*idum);
    for (j = NTAB+7; j >= 0; j--) {
      k = (*idum) / IQ1;
      *idum = IA1 * (*idum - k*IQ1) - k*IR1;
      if (*idum < 0) *idum += IM1;
      if (j < NTAB) iv[j] = *idum;
    }
    iy = iv[0];
  }

  k = (*idum) / IQ1;
  *idum = IA1 * (*idum - k*IQ1) - k*IR1;
  if (*idum < 0) *idum += IM1;
  k = idum2 / IQ2;
  idum2 = IA2 * (idum2 - k*IQ2) - k*IR2;
  if (idum2 < 0) idum2 += IM2;
  j = iy / NDIV;
  iy = iv[j] - idum2;
  iv[j] = *idum;
  if (iy < 1) iy += IMM1;
  if ((temp = AM * iy) > RNMX) return RNMX;
  else return temp;
}
